File: Src/Runtime/Strings/TrimLeft.cpp
/*
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #                           Typed                     # *
 * #                   A text formatting DSL             # *
 * #                                                     # *
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #         Created by Rodrigo R. & Contributors        # *
 * #         Released under the Apache License 2.0       # *
 * #            Check LICENSE.MD for more info           # *
 * #                                                     # *
 * #-----------------------------------------------------# *
*/

//
// Created by Rodrigo on 5/25/26.
//

#include "ADT/List/Object.h"
#include "TrimLeft.h"

#include "Support/Runtime/AccessString.h"

using namespace Typed;
using namespace Typed::Runtime;
using namespace Typed::Runtime::Strings;

ADT::Runtime::Object Strings::TrimLeft(
    ADT::List::Object &args,
    ADT::Lang::AST *trace
)
{
    auto &obj = args[0];
    auto str = Support::Runtime::AccessString(obj);
    const char *ptr = str.data();
    size_t size = str.size();
    size_t start = 0;

    // Remove all empty characters
    while (
        start < size &&
        (
            std::isblank(ptr[start]) ||
            std::iswblank(ptr[start]) ||
            std::isspace(ptr[start]) ||
            std::iswspace(ptr[start]) ||
            std::iscntrl(ptr[start]) ||
            std::iswcntrl(ptr[start])
        )
    )
    {
        ++start;
    }

    // Edge case: the string is empty
    if (start == size)
    {
        if (start > 0)
        {
            return {
                ADT::Runtime::ObjectType::String,
                std::string_view{}
            };
        }

        return obj;
    }

    // If the string is a view, return a view
    if (obj.type == ADT::Runtime::ObjectType::String)
    {
        return {
            ADT::Runtime::ObjectType::String,
            std::string_view(ptr + start, size - start)
        };
    }

    // Otherwise, return an owned string
    return {
        ADT::Runtime::ObjectType::String,
        std::string_view(ptr + start, size - start)
    };
}

File: src/runtime/strings/trim_left.rs
/*
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #                           Typed                     # *
 * #                   A text formatting DSL             # *
 * #                                                     # *
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #         Created by Rodrigo R. & Contributors        # *
 * #         Released under the Apache License 2.0       # *
 * #            Check LICENSE.MD for more info           # *
 * #                                                     # *
 * #-----------------------------------------------------# *
*/
use crate::adt::lang::{RuntimeArguments, AST};
use crate::adt::result::ExecutionResult;
use crate::adt::runtime::{HashableObject, Object, StringKind};
use crate::support::runtime::object::get_string;

pub fn trim_left<'a>(
    args: RuntimeArguments<'a>,
    trace: &AST<'a>
)
    -> ExecutionResult<'a>
{
    let origin_obj = args.get(0).unwrap();
    let origin = get_string(origin_obj, trace)?;
    
    // If the string is static, we can trim it without allocating
    if let Object::Hashable(
        HashableObject::String(
            StringKind::Static(s)
        )
    ) = origin_obj {
        let trimmed = s.trim_start();
        return Ok(
            Object::Hashable(
                HashableObject::String(
                    StringKind::Static(trimmed)
                )
            )
        );
    }
    
    // Otherwise, we need to allocate a new string
    let trimmed = String::from(origin.trim_start());
    
    Ok(
        Object::Hashable(
            HashableObject::String(
                StringKind::Dynamic(trimmed)
            )
        )
    )
}